gqlalchemy/memgraph.py

```diff
@@ -566,14 +569,67 @@ def __str__(self) -> str:
     def to_cypher_lambda(expression: str) -> str:
         """Method for creating a general lambda expression.
 
-        Variables `e` and `v` stand for relationship and node. The expression is
+        Variables `r` and `n` stand for relationship and node. The expression is
         used e.g. for a filter lambda, to use only relationships of length less
         than 200:
             expression="r.length < 200"
         with the filter lambda being:
             (r, n | r.length < 200)
 
         Args:
-            expression: Lambda conditions or statements.
+            expression: lambda conditions or statements.
         """
         return "" if expression is None else f"(r, n | {expression})"
+
+
+class BreadthFirstSearch(IntegratedAlgorithm):
+    """Build a BFS call for a Cypher query.
+
+    The Breadth-first search can be called in Memgraph with Cypher queries such
+    as: `MATCH (a {id: 723})-[*BFS ..10 (r, n | r.x > 12 AND n.y < 3)]-() RETURN *;`
+    It is called inside the relationship clause, `*BFS` naming the algorithm,
+    `..10` specifying depth bounds, and `(r, n | <expression>)` is a filter
+    lambda.
+    """
+
+    def __init__(
+        self,
+        lower_bound: int = None,
+        upper_bound: int = None,
+        condition: str = None,
+    ) -> None:
+        """
+        Args:
+            lower_bound: Lower bound for path depth. Defaults to `None`.
+            upper_bound: Upper bound for path depth. Defaults to `None`.
+            condition: Filter through nodes and relationships that pass this
+            condition. Defaults to `None`.
+        """
+        super().__init__()
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.condition = condition
+
+    def __str__(self) -> str:
+        """Get a Cypher query string for this algorithm."""
+        algo_str = BFS_EXPANSION
+
+        bounds = self.to_cypher_bounds()
+        if bounds != "":
+            algo_str = f"{algo_str} {bounds}"
+
+        filter_lambda = super().to_cypher_lambda(self.condition)
+        if filter_lambda != "":
+            algo_str = f"{algo_str} {filter_lambda}"
+
+        return algo_str
+
+    def to_cypher_bounds(self) -> str:
+        """If bounds are specified, returns them in grammar-defined form."""
+        if self.lower_bound is None and self.upper_bound is None:
+            return ""
+
+        lower = str(self.lower_bound) if self.lower_bound is not None else ""
+        upper = str(self.upper_bound) if self.upper_bound is not None else ""
```

Comment:
Move this to ctor maybe? It is overkill to make these checks and always convert None to "". Since this None value for bounds is not used anywhere else, set it as empty string in ctor if the passed parameter was None, and we can skip these conversions